### PR TITLE
[Community]: Added Simple Exceptions for PyPDFParser

### PR DESCRIPTION
**Pull request by keenborder786**
Issue: #24439
Added Simple Exceptions in `PyPDFParser`

### CODE REVIEW

File: libs/community/langchain_community/document_loaders/parsers/pdf.py [hunk 1]
```diff
@@ -119,16 +119,21 @@ def _extract_text_from_page(page: "PageObject") -> str:
...
+            try:
+                yield from [
+                    Document(
+                        page_content=_extract_text_from_page(page=page)
+                        + self._extract_images_from_page(page),
+                        metadata={"source": blob.source, "page": page_number},  # type: ignore[attr-defined]
+                    )
+                    for page_number, page in enumerate(pdf_reader.pages)
+                ]
+            except (pypdf.errors.PyPdfError, RecursionError) as e:
```
Comment (Avgor46) [inline libs/community/langchain_community/document_loaders/parsers/pdf.py:135]: Consider this https://github.com/py-pdf/pypdf/issues/2761. There are some other exceptions that can be thrown by pypdf library. Maybe you should catch any Exception here, or wait until issue is closed in pypdf.

Comment (keenborder786) [reply libs/community/langchain_community/document_loaders/parsers/pdf.py:135]: Sure I will do add other exceptions as well.

**Comment by keenborder786** [comment]
@ccurme please review

**Review by ccurme — COMMENTED** [review]
This PR catches errors and either:

1. Raises the same error again;
2. Raises an error of a different type.

What is the motivation for this change?

**Comment by keenborder786** [comment]
> This PR catches errors and either:
> 
> 1. Raises the same error again;
> 2. Raises an error of a different type.
> 
> What is the motivation for this change?

@ccurme motivation of this change is to explicitly make the aware of the exception being raised which is not being happening currently.

File: libs/community/langchain_community/document_loaders/parsers/pdf.py [hunk 2]
```diff
@@ -119,16 +119,29 @@ def _extract_text_from_page(page: "PageObject") -> str:
...
-                    page_content=_extract_text_from_page(page=page)
-                    + self._extract_images_from_page(page),
-                    metadata={"source": blob.source, "page": page_number},  # type: ignore[attr-defined]
-                )
-                for page_number, page in enumerate(pdf_reader.pages)
-            ]
+            try:
+                pdf_reader = pypdf.PdfReader(pdf_file_obj, password=self.password)
+            except pypdf.errors.PyPdfError as e:
+                raise e
```
Comment (ccurme) [inline libs/community/langchain_community/document_loaders/parsers/pdf.py:127]: This appears to have no effect as we raise the original error with no further action.

File: libs/community/langchain_community/document_loaders/parsers/pdf.py [hunk 3]
```diff
@@ -119,16 +119,29 @@ def _extract_text_from_page(page: "PageObject") -> str:
...
+                    for page_number, page in enumerate(pdf_reader.pages)
+                ]
+            except (
+                pypdf.errors.PyPdfError,
+                pypdf.errors.EmptyFileError,
+                pypdf.errors.PdfReadError,
+                pypdf.errors.PdfStreamError,
+                RecursionError,
+            ) as e:
+                raise ValueError(f"Your PDF could not be read due to error {e}")
```
[the remark this answers is not in the document]
Comment (keenborder786) [reply libs/community/langchain_community/document_loaders/parsers/pdf.py:144]: Yes that is intent to make make it more verbose.

[the remark this answers is not in the document]
Comment (keenborder786) [reply libs/community/langchain_community/document_loaders/parsers/pdf.py]: @ccurme please see now

**Comment by keenborder786** [comment]
@ccurme

**Comment by ccurme** [comment]
I'm not sure collapsing all the error types together into such a broad class (ValueError) is helpful. PyPDF implements a [PyPdfError](https://github.com/py-pdf/pypdf/blob/b7b3c8cedfc94d1b65fe2cd15741209b532e45c8/pypdf/errors.py#L19) which is the base class for all exceptions raised by PyPDF. If desired, you can catch that. Meanwhile, changing to ValueError loses our ability to catch specific errors.

**Pull request closed without merging**